feat: setup instructor dashboard filter

Ian2012 opened the pull request:
### Description

This PR adds a model for email preferences, an API to update that record, and the integration to the instructor dashboard to handle the preferences:

![image](https://github.com/eduNEXT/platform-plugin-forum-email-notifier/assets/33465240/166923f8-1d52-4828-a7fd-b5b74875ac5e)

### Testing instructions
In your environment, configure the following settings. The following plugin is used as reference:

```yaml
name: settings
version: 0.1.0
patches:
    openedx-common-settings: |
      OPEN_EDX_FILTERS_CONFIG = {
        "org.openedx.learning.instructor.dashboard.render.started.v1": {
          "fail_silently": False,
          "pipeline": [
            "platform_plugin_forum_email_notifier.extensions.filters.AddInstructorNotifierTab",
          ]
        },
      }
```
- Go to the instructor dashboard
- Go to Forum notification tab
- Select some of the available options
- Go to the admin and search for an stored record associated with that user, and that course for the preference you choose http://local.overhang.io:8000/admin/platform_plugin_forum_email_notifier

**Merge checklist:**
Check off if complete *or* not applicable:
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Fixup commits are squashed away
- [ ] Unit tests added/updated
- [ ] Manual testing instructions provided
- [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

Ian2012 commented:
@mariajgrimaldi how can I get the current user in the filter?

I need to inject the current preference of the user

mariajgrimaldi commented:
@Ian2012: can't you get the user from the current request? You can also inspect the instructor template context, but I'm 90% sure it doesn't include the logged in user.

mariajgrimaldi commented on platform_plugin_forum_email_notifier/admin.py:
```diff
@@ -0,0 +1,18 @@
+from django.contrib import admin
```
don't forget the module docstring

Ian2012 replied:
fixed

mariajgrimaldi commented on platform_plugin_forum_email_notifier/apps.py:
```diff
@@ -3,6 +3,7 @@
 """
 
 from django.apps import AppConfig
+from openedx.core.constants import COURSE_ID_PATTERN
```
We'd probably need to add a try-except

Ian2012 replied:
Fixed

mariajgrimaldi commented:
@Ian2012: can you add the test instructions? Thanks!

Ian2012 commented:
@mariajgrimaldi updated

mariajgrimaldi commented on platform_plugin_forum_email_notifier/models.py:
```diff
@@ -1,24 +1,53 @@
+
+class PreferenceOptions(models.IntegerChoices):
+    """
+    A model to store forum email notification preferences for a user.
```
Not exactly. What about `Preference options for a user related to email notifications frequency`?

mariajgrimaldi commented on platform_plugin_forum_email_notifier/tasks.py:
```diff
@@ -63,3 +65,45 @@ def send_email_notification(
+            ):
+                continue
+        except User.DoesNotExist:
+            log.warning(f"User {subscriber} does not exist")
```
we stop sending notifications if a user doesn't exist?

Ian2012 replied:
Yes, this shouldn't happen, but if it happens the task will finish succesfully

mariajgrimaldi replied:
why don't we ignore it and continue?

mariajgrimaldi replied:
what if a preference option does not exist for the user?

Ian2012 replied:
that's will cause a failure. I'll fix it

Ian2012 replied:
This is fixed in the email_digest branch

mariajgrimaldi commented on platform_plugin_forum_email_notifier/tasks.py:
```diff
@@ -63,3 +65,45 @@ def send_email_notification(
+    else:
+        raise ValueError(f"Invalid thread event type: {object_type}")
+
+    staff_subscribers = get_staff_subscribers(course_id)
```
what's the difference between getting subscribers and staff subscribers?

Ian2012 replied:
A subscriber is a user(staff or course staff) who is subscribed to a thread.

An staff subscriber is a course staff member that configured it's preference for the course to get notified of all posts

mariajgrimaldi commented on platform_plugin_forum_email_notifier/tasks.py:
```diff
@@ -63,3 +65,45 @@ def send_email_notification(
+            preference = ForumNotificationPreference.objects.get(
+                user=user, course_id=course_id
+            )
+            if preference.preference in (
+                PreferenceOptions.NONE,
+                PreferenceOptions.ALL_POSTS_DAILY_DIGEST,
+                PreferenceOptions.ALL_POSTS_WEEKLY_DIGEST,
+            ):
+                continue
```
if we're filtering for ALL_POSTS for staff subscribers, why are checking for these preferences as well? do non-staff users have preferences?

Ian2012 replied:
Not at the moment, we should do that only for staff users

Ian2012 replied:
This is fixed in the email digest branch

mariajgrimaldi commented on platform_plugin_forum_email_notifier/utils.py:
```diff
@@ -32,6 +32,26 @@ def get_subscribers(thread_id):
 
 
+def get_staff_subscribers(course_id):
+    """Return a list of user ids subscribed to a thread."""
```
can we be a bit more specific of what a staff subscriber is here in the docstring?

mariajgrimaldi commented on platform_plugin_forum_email_notifier/utils.py:
```diff
@@ -32,6 +32,26 @@ def get_subscribers(thread_id):
     return f"{settings.PREFIX}/threads/{thread_id}/subscriptions"
+
+
+class ForumObject(IntEnum):
```
can we add a docstring here

mariajgrimaldi commented:
Last few comments. Can we make the tests pass?

mariajgrimaldi reviewed (approved):
LGTM! We tested this in a remote environment :)!

Ian2012 committed:
feat: setup instructor dashboard integration

fix: latest corrections for quality

feat: setup instructor dashboard

feat: listen for forum events

feat: implement email template and celery task

feat: migrate to edx_ace

chore: install edx_ace

chore: refactor for quality

(cherry picked from commit f72eaebee31deb1b5b36390c0bc98515585f55cd)

chore: upgrade openedx-events

feat: setup instructor dashboard filter

chore: install openedx filters and web_fragments

feat: define model and view por forum email preference

feat: implement forum email preference logic

fix: show current preference in instructor dashboard

fix: add missing docstring for admin module

chore: add docstring for staff subcribers

(cherry picked from commit 5cc188d25a66b59f57e0784febb83c7db607b194)

chore: add docstring for forum object

fix: handle preference doesn't exists

fix: return context and template name for filter

fix: return context only

chore: install django rest framework

chore: refactor for quality

fix: run upgrade

fix: conflicts

fix: include .txt templates in distribution

BryanttV commented:
Tested locally and it works!
